komodo/pypi_dependencies.py

```diff
@@ -0,0 +1,234 @@
+import os
+import platform
+import subprocess
+import sys
+from tempfile import TemporaryDirectory
+from typing import Dict, List, Set, Tuple
+
+import pkginfo
+import yaml
+from packaging.requirements import Requirement
+from packaging.utils import canonicalize_name
+from packaging.version import InvalidVersion
+
+from .package_version import LATEST_PACKAGE_ALIAS
+
+
+# From Pep 508
+def format_full_version(info) -> str:
+    version = f"{info.major}.{info.minor}.{info.micro}"
+    kind = info.releaselevel
+    if kind != "final":
+        version += kind[0] + str(info.serial)
+    return version
+
+
+environment = {
+    "implementation_name": sys.implementation.name,
+    "implementation_version": format_full_version(sys.implementation.version),
+    "os_name": os.name,
+    "platform_machine": platform.machine(),
+    "platform_python_implementation": platform.python_implementation(),
+    "platform_release": platform.release(),
+    "platform_system": platform.system(),
+    "platform_version": platform.version(),
+    "python_full_version": platform.python_version(),
+    "python_version": ".".join(platform.python_version_tuple()[:2]),
+    "sys_platform": sys.platform,
+}
+
+
+# pylint: disable=too-many-instance-attributes
+class PypiDependencies:
+    def __init__(
+        self,
+        to_install: Dict[str, str],
+        python_version: str,
+        cachefile: str = "./pypi_dependencies.yml",
+    ) -> None:
+        self.python_version = python_version
+        environment["python_full_version"] = python_version
+        environment["python_version"] = python_version[: python_version.rindex(".")]
+        self.satisfied_requirements = set()
+        self.failed_requirements = set()
+        self.used_packages = set()
+
+        self._to_install = to_install
+
+        self._install_names = {canonicalize_name(name): name for name in to_install}
+        self._to_install = {
+            canonicalize_name(name): version for name, version in to_install.items()
+        }
+        self._cachefile = cachefile
+        if os.path.exists(self._cachefile):
+            with open(self._cachefile, "r", encoding="utf-8") as f:
+                self.requirements = yaml.safe_load(f)
+                self.requirements = {
+                    canonicalize_name(name): r for name, r in self.requirements.items()
+                }
+        else:
+            self.requirements = {}
+
+        self._user_specified = {}
+
+    def check(self, packages=None):
+        for package_name, version in packages or self._to_install.items():
+            requirements = self.get_requirements(package_name, version)
+            self.used_packages.add(package_name)
+            for r in requirements:
+                _ = self.satisfied(r)
```

Comment:
Do we `check` here? Aren't we collecting requirements rather?
If satisfied fails, do we notice?

Reply:
I guess the name is confusing. Let me try to fix it